metadetect/lsst/metadetect.py

```diff
@@ -127,147 +115,203 @@ def detect_deblend_and_measure(
-                mbexp=mbexp,
-                detexp=detexp,
-                sources=sources,
-                fitter=fitter,
-                stamp_size=config['stamp_size'],
-                Tvals=Tvals,
-                shredder_config=shredder_config,
-                rng=rng,
-                show=show,
-            )
-
-    else:
-
-        LOG.info('measuring with blended stamps')
-
-        mbexp = util.get_mbexp(exposures)
-
-        for obslist in mbobs:
-            assert len(obslist) == 1, 'no multiepoch'
-
-        sources, detexp = measure.detect_and_deblend(
-            mbexp=mbexp,
-            rng=rng,
-            thresh=config['detect']['thresh'],
-            show=show,
-        )
-        results = measure.measure(
-            mbexp=mbexp,
-            detexp=detexp,
-            sources=sources,
-            fitter=fitter,
-            stamp_size=config['stamp_size'],
-            find_cen=config['find_cen'],
-            rng=rng,  # needed if find_cen is True
-        )
+    sources, detexp = measure.detect_and_deblend(
+        mbexp=mbexp,
+        rng=rng,
+        thresh=config['detect']['thresh'],
+        show=show,
+    )
+    results = measure.measure(
+        mbexp=mbexp,
+        detexp=detexp,
+        sources=sources,
+        fitter=fitter,
+        stamp_size=config['stamp_size'],
+    )
 
     return results
 
 
-def add_noshear_pos(config, res, shear_str, obs):
+def add_noshear_pos_exp(res, shear_str, exp):
     """
     add unsheared positions to the input result array
     """
+
+    dims = exp.image.array.shape
+
+    cen, _ = get_integer_center(
+        wcs=exp.getWcs(),
+        bbox=exp.getBBox(),
+        as_double=True,
+    )
+    jac = get_jacobian(exp=exp, cen=cen)
+
     rows_noshear, cols_noshear = shearpos.unshear_positions(
         res['row'] - res['row0'],
         res['col'] - res['col0'],
         shear_str,
-        obs,  # an example for jacobian and image shape
+        jac=jac,
+        dims=dims,
     )
     res['row_noshear'] = rows_noshear
     res['col_noshear'] = cols_noshear
 
 
-def add_mfrac(config, mfrac, res, obs):
+def add_mfrac(config, mfrac, res, exp):
     """
     calculate and add mfrac to the input result array
     """
     if np.any(mfrac > 0):
+
         # we are using the positions with the metacal shear removed for
         # this.
-        res['mfrac'] = measure_mfrac(
+
+        cen, _ = get_integer_center(
+            wcs=exp.getWcs(),
+            bbox=exp.getBBox(),
+            as_double=True,
+        )
+        jac = get_jacobian(exp=exp, cen=cen)
+
+        res['mfrac'] = measure_weighted_mfrac(
             mfrac=mfrac,
             x=res['col_noshear'],
             y=res['row_noshear'],
-            box_sizes=res['stamp_size'],
-            obs=obs,
+            jac=jac,
             fwhm=config.get('mfrac_fwhm', None),
         )
     else:
         res['mfrac'] = 0
 
 
+def measure_weighted_mfrac(
```

Comment:
This is duplicated code. It could be moved up and out and refactored like you did shearpos before.

Reply:
I thought about it, but it only shares a few lines of code with the other one

Reply:
kk